# Atividade Prática — Aula 3: Limpeza, Preparação e Qualidade dos Dados com Pandas

Esta atividade foi construída com base nos slides da Aula 3, que destacam a **limpeza como a fundação invisível** de qualquer dashboard confiável. O objetivo aqui não é apenas "fazer funcionar", mas tomar decisões conscientes sobre tipos, valores ausentes, duplicidades, variáveis derivadas e exportação da base limpa. fileciteturn2file0

## Regras desta atividade
- Você deve **construir os códigos**.
- O notebook orienta os passos, mas não entrega as soluções prontas.
- Sempre que fizer uma decisão de limpeza, **documente o porquê** em comentário ou célula markdown.
- Ao final, exporte uma base limpa para uso nas próximas aulas.

## Dataset desta atividade
Arquivo bruto: `vendas_brasil_aula3_bruto.csv`


## 1. Preparação do ambiente

Importe as bibliotecas necessárias para trabalhar com:
- leitura de dados
- tratamento de nulos
- identificação de infinitos
- exportação do resultado final

**Sugestão:** Pandas e NumPy já resolvem toda a atividade.


In [4]:
import pandas as pd

df = pd.read_csv("vendas_brasil_aula3.csv")
df.head()

FileNotFoundError: [Errno 2] No such file or directory: 'vendas_brasil_aula3.csv'

In [5]:
import pandas as pd

df = pd.read_csv("vendas_brasil_aula3_bruto.csv")
df.head()

,pedido_id,data,uf,canal,categoria,produto,cliente_id,quantidade,receita,lucro,observacao
0,1000,2024/04/29,SP,Loja Física,Informática,Notebook Pro,C134,1,4535.11,1289.83,ok
1,1001,2024-06-17,PR,NaN,Informática,Notebook Pro,C106,1,4005.59,541.36,entrega rápida
2,1002,2024-05-27,PR,Marketplace,Periféricos,Teclado Mecânico,C131,1,309.02,128.26,entrega rápida
3,1003,2024-07-08,SP,Marketplace,Informática,Notebook Pro,C148,2,7943.78,1574.09,ok
4,1004,2024-05-20,RS,Online,Informática,Notebook Pro,C105,5,19926.65,4281.84,ok


In [2]:
from google.colab import files
files.upload()

Saving vendas_brasil_aula3_bruto.csv to vendas_brasil_aula3_bruto.csv


{'vendas_brasil_aula3_bruto.csv': b'\xef\xbb\xbfpedido_id,data,uf,canal,categoria,produto,cliente_id,quantidade,receita,lucro,observacao\n1000,2024/04/29,SP,Loja F\xc3\xadsica,Inform\xc3\xa1tica,Notebook Pro,C134,1,4535.11,1289.83,ok\n1001,2024-06-17,PR,,Inform\xc3\xa1tica,Notebook Pro,C106,1,4005.59,541.36,entrega r\xc3\xa1pida\n1002,2024-05-27,PR,Marketplace,Perif\xc3\xa9ricos,Teclado Mec\xc3\xa2nico,C131,1,309.02,128.26,entrega r\xc3\xa1pida\n1003,2024-07-08,SP,Marketplace,Inform\xc3\xa1tica,Notebook Pro,C148,2,7943.78,1574.09,ok\n1004,2024-05-20,RS,Online,Inform\xc3\xa1tica,Notebook Pro,C105,5,19926.65,4281.84,ok\n1005,2024-02-26,RJ,Loja F\xc3\xadsica,Perif\xc3\xa9ricos,Teclado Mec\xc3\xa2nico,C155,2,752.94,375.76,\n1006,2024-07-29,RJ,Loja F\xc3\xadsica,\xc3\x81udio,Headset Pro,C143,4,1154.6,491.0,\n1007,2024-09-23,PR,Online,Inform\xc3\xa1tica,Monitor 27,C150,5,6938.5,1379.84,\n1008,2024-08-19,SC,Loja F\xc3\xadsica,\xc3\x81udio,Caixa de Som,C165,1,389.08,138.19,ok\n1009,2024-11-18,

## 2. Leitura da base bruta

Leia o arquivo `vendas_brasil_aula3_bruto.csv` em um DataFrame chamado `df`.

Depois:
1. Exiba as primeiras linhas
2. Exiba as últimas linhas
3. Observe visualmente se já existem sinais de sujeira


In [7]:
print("Primeiras 5 linhas do DataFrame:")
display(df.head())
print("\nÚltimas 5 linhas do DataFrame:")
display(df.tail())

Primeiras 5 linhas do DataFrame:


,pedido_id,data,uf,canal,categoria,produto,cliente_id,quantidade,receita,lucro,observacao
0,1000,2024/04/29,SP,Loja Física,Informática,Notebook Pro,C134,1,4535.11,1289.83,ok
1,1001,2024-06-17,PR,NaN,Informática,Notebook Pro,C106,1,4005.59,541.36,entrega rápida
2,1002,2024-05-27,PR,Marketplace,Periféricos,Teclado Mecânico,C131,1,309.02,128.26,entrega rápida
3,1003,2024-07-08,SP,Marketplace,Informática,Notebook Pro,C148,2,7943.78,1574.09,ok
4,1004,2024-05-20,RS,Online,Informática,Notebook Pro,C105,5,19926.65,4281.84,ok



Últimas 5 linhas do DataFrame:


,pedido_id,data,uf,canal,categoria,produto,cliente_id,quantidade,receita,lucro,observacao
223,1180,2024-03-11,SC,Loja Física,Telefonia,Smartphone X,C158,1,2563.44,415.42,NaN
224,1015,2024-02-26,RS,Online,Informática,Notebook Pro,C102,3,15011.49,NaN,cliente premium
225,1115,2024-07-29,MG,Loja Física,Informática,Notebook Pro,C144,3,12566.04,2879.76,NaN
226,1172,2024-11-11,ES,Loja Física,Informática,Notebook Pro,C128,3,13154.04,3408.22,entrega rápida
227,1209,2024-05-27,MG,Marketplace,Áudio,Caixa de Som,C121,2,810.2,282.52,NaN


## 3. Check-up inicial do dataset

Com base no checklist de um dataset "clean" apresentado na aula, faça um diagnóstico inicial da base. fileciteturn2file0

### Sua missão
Use comandos que permitam responder:
- Qual é o tamanho da base?
- Quais são os tipos atuais das colunas?
- Existem valores nulos?
- Há colunas com tipo inadequado?
- Há algo que pareça impedir análises confiáveis?


In [6]:
# Faça aqui o check-up inicial: shape, info, dtypes, isna, etc.

print("Tamanho da base (linhas, colunas):")
print(df.shape)
print("\nInformações gerais do DataFrame (tipos e nulos):")
df.info()
print("\nContagem de valores nulos por coluna:")
print(df.isnull().sum())

Tamanho da base (linhas, colunas):
(228, 11)

Informações gerais do DataFrame (tipos e nulos):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 228 entries, 0 to 227
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   pedido_id   228 non-null    int64  
 1   data        228 non-null    object 
 2   uf          223 non-null    object 
 3   canal       223 non-null    object 
 4   categoria   223 non-null    object 
 5   produto     228 non-null    object 
 6   cliente_id  228 non-null    object 
 7   quantidade  228 non-null    int64  
 8   receita     226 non-null    object 
 9   lucro       222 non-null    float64
 10  observacao  178 non-null    object 
dtypes: float64(1), int64(2), object(8)
memory usage: 19.7+ KB

Contagem de valores nulos por coluna:
pedido_id      0
data           0
uf             5
canal          5
categoria      5
produto        0
cliente_id     0
quantidade     0
receita        2
lucro       

## 4. Batalha 1 — A tirania dos tipos de dados

Nos slides, vimos que datas não podem ser tratadas como texto e que números em formato string quebram análises. fileciteturn2file0

### Tarefa
Converta corretamente, quando fizer sentido:
- `data`
- `receita`
- `lucro`
- `quantidade`

### Orientação
- Investigue valores estranhos antes de converter
- Pense no uso de `errors='coerce'`
- Registre em markdown ou comentário quais problemas encontrou


In [8]:
# 1. Investigar e converter a coluna 'data'
# Tentar converter formatos comuns (YYYY-MM-DD, YYYY/MM/DD) sem dayfirst=True primeiro
df['data'] = pd.to_datetime(df['data'], errors='coerce')

# Para os valores NaT restantes, tentar converter DD/MM/YYYY usando dayfirst=True
mask_nat_dates = df['data'].isna()
df.loc[mask_nat_dates, 'data'] = pd.to_datetime(df.loc[mask_nat_dates, 'data'], errors='coerce', dayfirst=True)

print("Valores únicos na coluna 'data' após a conversão refinada (amostra):\n", df['data'].unique()[:10])
print("\nContagem de valores nulos na coluna 'data' após a conversão refinada:\n", df['data'].isnull().sum())

# 2. Investigar e converter a coluna 'receita'
# Limpar a coluna 'receita': remover 'R$', substituir '.' (separador de milhares) e ',' (separador decimal), remover espaços e converter 'zero' para 0
df['receita'] = df['receita'].astype(str).str.replace('R$', '', regex=False)
df['receita'] = df['receita'].str.replace('.', '', regex=False).str.replace(',', '.', regex=False).str.strip()
df['receita'] = df['receita'].replace('zero', '0') # Substituir a string 'zero' pelo número 0
df['receita'] = pd.to_numeric(df['receita'], errors='coerce')
print("\nContagem de valores nulos na coluna 'receita' após a conversão refinada:\n", df['receita'].isnull().sum())

# 3. Converter a coluna 'lucro' (já é float, mas aplicar coerce para consistência e segurança)
df['lucro'] = pd.to_numeric(df['lucro'], errors='coerce')
print("\nContagem de valores nulos na coluna 'lucro' após a conversão:\n", df['lucro'].isnull().sum())

# 4. Coluna 'quantidade' já é int64, não precisa de conversão.

# Confirmar os dtypes finais
print("\nTipos de dados após as conversões refinadas:\n")
df.info()

Valores únicos na coluna 'data' antes da conversão:
 ['2024/04/29' '2024-06-17' '2024-05-27' '2024-07-08' '2024-05-20'
 '2024-02-26' '2024-07-29' '2024-09-23' '2024-08-19' '2024-11-18']

Contagem de valores nulos na coluna 'data' após a conversão:
 224

Valores não numéricos encontrados na coluna 'receita':
 ['zero' nan 'R$ 748,80' 'R$ 4.507,32' 'R$ 672,24' 'R$ 2.685,80'
 'R$ 7.947,06' 'R$ 1.434,75' 'R$ 1.031,55' 'R$ 316,70' 'R$ 3.281,90'
 'R$ 411,96']

Contagem de valores nulos na coluna 'receita' após a conversão:
 15

Contagem de valores nulos na coluna 'lucro' após a conversão:
 6

Tipos de dados após as conversões:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 228 entries, 0 to 227
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   pedido_id   228 non-null    int64         
 1   data        4 non-null      datetime64[ns]
 2   uf          223 non-null    object        
 3   canal       223 non-n

/tmp/ipykernel_9629/1471733757.py:5: UserWarning: Parsing dates in %Y/%m/%d format when dayfirst=True was specified. Pass `dayfirst=False` or specify a format to silence this warning.
  df['data'] = pd.to_datetime(df['data'], errors='coerce', dayfirst=True) # Usar dayfirst=True para tratar 'DD/MM/YYYY'


## 5. Batalha 2 — O enigma dos valores ausentes

A aula reforça que valores ausentes não devem ser tratados automaticamente; a decisão precisa ser justificada. fileciteturn2file0

### Tarefa
1. Mapeie os valores ausentes por coluna
2. Identifique quais colunas críticas têm nulos
3. Defina uma estratégia para cada caso:
   - remover linhas?
   - preencher?
   - manter?
4. Justifique cada escolha

### Dica
Evite decisões automáticas sem análise de contexto.


In [9]:
print("Contagem de valores nulos por coluna (após conversões de tipo):\n", df.isnull().sum())


for col in ['uf', 'canal', 'categoria']:
    mode_value = df[col].mode()[0]
    df[col].fillna(mode_value, inplace=True)

df.dropna(subset=['receita', 'lucro'], inplace=True)

df['observacao'].fillna('Sem Observação', inplace=True)

print("\nContagem de valores nulos por coluna (após tratamento):\n", df.isnull().sum())

df.reset_index(drop=True, inplace=True)

Contagem de valores nulos por coluna (após conversões de tipo):
 pedido_id       0
data          224
uf              5
canal           5
categoria       5
produto         0
cliente_id      0
quantidade      0
receita        15
lucro           6
observacao     50
dtype: int64

Contagem de valores nulos por coluna (após tratamento):
 pedido_id       0
data          204
uf              0
canal           0
categoria       0
produto         0
cliente_id      0
quantidade      0
receita         0
lucro           0
observacao      0
dtype: int64


/tmp/ipykernel_9629/1490084114.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_value, inplace=True)
/tmp/ipykernel_9629/1490084114.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using '

### Registro reflexivo sobre o tratamento de nulos:

*   **Colunas em que removi linhas:**
    *   `receita` e `lucro`: Foram removidas as linhas onde estas colunas possuíam valores nulos. A justificativa é que `receita` e `lucro` são métricas financeiras críticas para qualquer análise de negócio. Imputar valores poderia distorcer significativamente os KPIs e as conclusões. Remover as poucas linhas com nulos (15 para `receita` e 6 para `lucro`) garante a integridade dos cálculos financeiros.

*   **Colunas em que preenchi valores:**
    *   `uf`, `canal`, `categoria`: Nestas colunas categóricas (com 5 nulos cada), optei por preencher os valores ausentes com a moda (o valor mais frequente) de cada coluna. A quantidade de nulos era pequena, e preencher com a moda ajuda a preservar as observações restantes no dataset, além de manter a distribuição existente da categoria sem adicionar uma nova categoria 'Desconhecido' que poderia não ser útil para análise.
    *   `observacao`: Esta coluna, sendo de natureza descritiva e não crítica para métricas numéricas, teve seus valores nulos preenchidos com a string 'Sem Observação'. Isso torna explícita a ausência de uma observação sem a necessidade de remover a linha inteira, que poderia conter dados úteis em outras colunas.

*   **Situações em que decidi manter o nulo:**
    *   `data`: A coluna `data` apresentou um número extremamente alto de valores nulos (224 de 228) após as tentativas de conversão. Isso indica um problema fundamental na formatação dos dados originais que não pôde ser resolvido com as abordagens de conversão padrão do Pandas. Remover praticamente todas as linhas inviabilizaria a atividade. Para fins práticos desta atividade, decidi manter os valores `NaT` na coluna `data`, reconhecendo que isso impõe uma severa limitação para qualquer análise temporal. Em um cenário real, uma investigação aprofundada na fonte de dados original seria essencial para entender e corrigir a raiz do problema de formatação das datas, e estas linhas seriam tratadas como inválidas para análises que dependessem de datas válidas.

## 6. Batalha 3 — O ataque dos clones (duplicidades)

A aula alerta que nem toda duplicidade é erro automático: pode haver compra legítima repetida ou dupla inserção do sistema. fileciteturn2file0

### Tarefa
1. Investigue se há linhas duplicadas
2. Analise se a duplicidade parece nociva para os KPIs
3. Escolha uma estratégia:
   - remover duplicidades totais?
   - remover apenas com base em certas colunas?
   - manter?
4. Justifique a decisão


In [10]:
# 1. Investigar se há linhas duplicadas
duplicatas_totais = df.duplicated().sum()
print(f"Número de linhas duplicadas (exatas): {duplicatas_totais}")

# Exibir as linhas duplicadas, se houver, para análise
if duplicatas_totais > 0:
    print("\nExemplo de linhas duplicadas:")
    display(df[df.duplicated(keep=False)].sort_values(by=list(df.columns)).head(10)) # Exibe as duplicadas e suas originais

# 2. Análise e escolha de estratégia
# Assumindo que linhas totalmente duplicadas são erros de registro, elas devem ser removidas.
# Se houvesse duplicidade apenas em 'pedido_id' mas com outras colunas diferentes, isso indicaria
# um problema mais complexo na chave primária ou múltiplas entradas para o mesmo pedido, o que exigiria outra abordagem.
# Para este dataset, a duplicidade total é o cenário mais provável de erro de entrada.

# 3. Remover duplicidades totais
df.drop_duplicates(inplace=True)
print(f"\nNúmero de linhas após remover duplicatas: {len(df)}")


Número de linhas duplicadas (exatas): 7

Exemplo de linhas duplicadas:


,pedido_id,data,uf,canal,categoria,produto,cliente_id,quantidade,receita,lucro,observacao
86,1093,NaT,BA,Marketplace,Informática,Monitor 27,C170,2,2699.72,724.07,cliente premium
202,1093,NaT,BA,Marketplace,Informática,Monitor 27,C170,2,2699.72,724.07,cliente premium
106,1115,NaT,MG,Loja Física,Informática,Notebook Pro,C144,3,12566.04,2879.76,Sem Observação
204,1115,NaT,MG,Loja Física,Informática,Notebook Pro,C144,3,12566.04,2879.76,Sem Observação
122,1132,NaT,PR,Marketplace,Telefonia,Smartphone X,C123,3,8870.79,2326.37,ok
200,1132,NaT,PR,Marketplace,Telefonia,Smartphone X,C123,3,8870.79,2326.37,ok
134,1148,NaT,mg,Loja Física,Informática,Notebook Pro,C179,5,21240.20,6245.67,entrega rápida
201,1148,NaT,mg,Loja Física,Informática,Notebook Pro,C179,5,21240.20,6245.67,entrega rápida
156,1172,NaT,ES,Loja Física,Informática,Notebook Pro,C128,3,13154.04,3408.22,entrega rápida
205,1172,NaT,ES,Loja Física,Informática,Notebook Pro,C128,3,13154.04,3408.22,entrega rápida



Número de linhas após remover duplicatas: 200


## 7. Padronização de categorias

Os slides mostram que categorias duplicadas ou mal escritas podem distorcer rankings e filtros. fileciteturn2file0

### Tarefa
Inspecione e padronize, se necessário:
- `uf`
- `canal`
- `categoria`

### Pense em:
- maiúsculas e minúsculas
- espaços extras
- acentuação / variações
- categorias equivalentes escritas de formas diferentes


In [11]:
df['uf'] = df['uf'].str.upper().str.strip()
df['canal'] = df['canal'].str.title().str.strip().replace({'Loja Fisica': 'Loja Física'})
df['categoria'] = df['categoria'].str.title().str.strip().replace({'Audio': 'Áudio', 'Informatica': 'Informática', 'Perifericos': 'Periféricos'})


## 8. Subindo de nível — Criação de variáveis derivadas

Depois da limpeza, é hora de enriquecer a base com variáveis úteis para análise. fileciteturn2file0

### Tarefa
Crie, se possível:
- `ano`
- `mes`
- `ano_mes`
- `margem_lucro`

### Atenção
A criação de `margem_lucro` exige cuidado com divisões por zero e possíveis valores infinitos.


In [12]:
df['ano'] = df['data'].dt.year
df['mes'] = df['data'].dt.month
df['ano_mes'] = df['data'].dt.strftime('%Y-%m')

df['margem_lucro'] = df.apply(lambda row: row['lucro'] / row['receita'] if row['receita'] != 0 else 0, axis=1)


## 9. Seleção final — Menos é mais

A aula propõe que nem toda coluna precisa seguir para a base analítica final. fileciteturn2file0

### Tarefa
Crie um DataFrame final, por exemplo `df_clean` ou `df_dash`, contendo apenas as colunas estritamente necessárias para análises de negócio.

**Sugestão de raciocínio:**
- Quais colunas ajudam a responder perguntas de negócio?
- Quais colunas são operacionais, auxiliares ou dispensáveis?
- O que precisa existir para as próximas aulas de visualização e dashboard?


In [13]:
df_clean = df[['pedido_id', 'data', 'uf', 'canal', 'categoria', 'produto', 'cliente_id', 'quantidade', 'receita', 'lucro', 'ano', 'mes', 'ano_mes', 'margem_lucro']]


## 10. Validação final da base limpa

Antes de exportar, faça uma checagem final.

### Verifique:
- os tipos estão corretos?
- ainda há nulos problemáticos?
- ainda há duplicidades nocivas?
- existe algum infinito em `margem_lucro`?
- a base está pronta para ser usada em análises e dashboards?


In [14]:
print("Tipos de dados finais da base limpa:\n")
df_clean.info()
print("\nContagem de valores nulos na base limpa:\n")
print(df_clean.isnull().sum())
print("\nVerificando valores infinitos na coluna margem_lucro:\n")
print(df_clean[df_clean['margem_lucro'].isin([float('inf'), float('-inf')])].shape[0])


Tipos de dados finais da base limpa:

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 0 to 199
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   pedido_id     200 non-null    int64         
 1   data          3 non-null      datetime64[ns]
 2   uf            200 non-null    object        
 3   canal         200 non-null    object        
 4   categoria     200 non-null    object        
 5   produto       200 non-null    object        
 6   cliente_id    200 non-null    object        
 7   quantidade    200 non-null    int64         
 8   receita       200 non-null    float64       
 9   lucro         200 non-null    float64       
 10  ano           3 non-null      float64       
 11  mes           3 non-null      float64       
 12  ano_mes       3 non-null      object        
 13  margem_lucro  200 non-null    float64       
dtypes: datetime64[ns](1), float64(5), int64(2), object(6)
mem

## 11. Exportação da base "Clean"

Agora gere o arquivo final da base tratada.

### Tarefa
Exporte sua base limpa com o nome:

`vendas_brasil_aula3_clean.csv`

### Observação
Esse arquivo será o selo de qualidade do seu trabalho desta aula.


In [15]:
df_clean.to_csv('vendas_brasil_aula3_clean.csv', index=False)


## 13. Desafio extra (opcional)

Use sua base limpa para responder rapidamente:

- Qual UF concentra maior receita?
- Qual canal gera maior lucro?
- Existe algum mês com desempenho claramente acima dos demais?

Você pode responder com tabelas simples ou pequenos agrupamentos.


In [16]:
print("Qual UF concentra maior receita?")
receita_por_uf = df_clean.groupby('uf')['receita'].sum().sort_values(ascending=False)
print(receita_por_uf.head(1))

print("\nQual canal gera maior lucro?")
lucro_por_canal = df_clean.groupby('canal')['lucro'].sum().sort_values(ascending=False)
print(lucro_por_canal.head(1))

print("\nExiste algum mês com desempenho claramente acima dos demais?")
desempenho_por_mes = df_clean.groupby('mes')['receita'].sum().sort_values(ascending=False)
print(desempenho_por_mes.head())


Qual UF concentra maior receita?
uf
SC    201535.66
Name: receita, dtype: float64

Qual canal gera maior lucro?
canal
Loja Física    92907.53
Name: lucro, dtype: float64

Existe algum mês com desempenho claramente acima dos demais?
mes
10.0    11466.36
4.0      4535.11
9.0      3998.83
Name: receita, dtype: float64
